VNET/VXLAN changes

prsunny opened the pull request:
**What I did**
Modified Vxlanorch to support VNet peering feature.
Introduced the following:
1. Vxlan VRF mapper
2. Generic code to handle BRIDGE/VRF/VLAN encap/decap mapper
3. Test case update for existing Vxlan VS test. 

**Why I did it**
Support VNet peering using Vxlan (VNI <-> VRF)

**How I verified it**
Verified by VS test and checking the ASIC DB for intended SAI tables

**Details if related**
Design Document:
https://github.com/prsunny/SONiC/blob/gh-pages/doc/vxlan/Vxlan_hld.md
Related PR
https://github.com/Azure/sonic-swss/pull/632

Pending:
1. VS test case for VNET (new code)
2. Delete sequence.

prsunny commented:
retest this please

qiluo-msft commented:
retest this please

prsunny commented:
retest this please

prsunny commented:
retest this please

lguohan commented:
@prsunny , can you rebase your pr and do a test again? I fixed a few things on the vstest.

lguohan commented:
retest this please

lguohan commented on orchagent/vnetorch.h:
```diff
@@ -192,6 +194,13 @@ class VNetRouteRequest : public Request
 using NextHopMap = map<IpAddress, sai_object_id_t>;
 using NextHopTunnels = map<string, NextHopMap>;
 
+struct endpoint
```
rename to tunnelEndpoint for clarity?

lguohan commented:
is the vs test in this PR or a new PR?

lguohan commented on orchagent/vxlanorch.cpp:
```diff
@@ -312,3 +547,135 @@ bool VxlanTunnelMapOrch::delOperation(const Request& request)
+    if (vni_id >= 1<<24)
+    {
+        SWSS_LOG_ERROR("Vxlan vni id is too big: %d", vni_id);
+        return true;
```
>true [](start = 15, length = 4)

why return true here?

prsunny replied:
Returning true to remove the entry from msgq to avoid further retries.

lguohan commented on orchagent/vxlanorch.cpp:
```diff
@@ -312,3 +547,135 @@ bool VxlanTunnelMapOrch::delOperation(const Request& request)
+    const auto full_map_entry_name = request.getFullKey();
+    if (isVrfMapExists(full_map_entry_name))
+    {
+        SWSS_LOG_ERROR("Vxlan map '%s' is already exist", full_map_entry_name.c_str());
```
>SWSS_LOG_ERROR [](start = 8, length = 14)

is this ERROR, it should be just a NOTICE.

lguohan commented on orchagent/vxlanorch.cpp:
```diff
@@ -312,3 +547,135 @@ bool VxlanTunnelMapOrch::delOperation(const Request& request)
+    }
+    else if (vrf_orch->isVRFexists(vrf_name))
+    {
+        //Do VRF task
```
>//Do VRF task [](start = 8, length = 13)

is this a placeholder? what is do vrf task?

prsunny replied:
Yes, it is a placeholder for VRF non-vnet usecase. I've updated the comment.

lguohan commented on tests/test_vnet.py:
```diff
@@ -0,0 +1,705 @@
+
+        create_vxlan_tunnel(dvs, tunnel_name, '10.10.10.10')
+
+        vr_ids = create_vnet_entry(dvs, 'Vnet_2000', tunnel_name, '2000', "", self.vnet_vr_ids)
```
can you create vr_ids as a map here? vnet_vrids={'ing':vrid, 'egr':vrid}? It will be easier to understand the test.

lguohan commented on tests/test_vnet.py:
```diff
@@ -0,0 +1,705 @@
+
+        create_vlan_interface(dvs, "Vlan101", self.vlan_ids, "Ethernet28", "Vnet_2000", "100.100.4.1/24", vr_ids[0])
+
+        get_vr_set(vr_set, [vr_ids])
```
it is not intuitive of this get_vr_set function. can you try to remove it and pass the vnet_vrids into create_vnet_routes()? there, you know whether it is ingress or egress route and you can check inside the create_vnet_routes()?

lguohan commented on tests/test_vnet.py:
```diff
@@ -0,0 +1,705 @@
+        create_vnet_routes(dvs, "100.100.1.1/32", 'Vnet_2000', '10.10.10.1', vr_set, tun_id, self.nh_ids)
+
+        get_vr_set(vr_set, [vr_ids], False)
+        create_vnet_if_routes(dvs, "100.100.3.0/24", 'Vnet_2000', 'Vlan100', vr_set)
```
is create_vnet_local_routes more intuitive?

pavel-shirshov commented on orchagent/vxlanorch.cpp:
```diff
@@ -47,8 +106,117 @@ create_tunnel_map()
+sai_status_t create_nexthop_tunnel(
+    sai_ip_address_t host_ip,
+    sai_uint32_t vni, // optional vni
+    sai_mac_t *mac, // inner destination mac
```
sai_mac_t is defined as `typedef UINT8   sai_mac_t[6];`

pavel-shirshov commented on orchagent/vxlanorch.cpp:
```diff
@@ -124,18 +313,17 @@ create_tunnel(sai_object_id_t tunnel_map_id)
     sai_object_id_t tunnel_oid,
-    sai_ip4_t srcip,
-    sai_ip4_t dstip,
+    sai_ip_address_t *srcip,
```
do we need srcip as pointer here?

pavel-shirshov commented on orchagent/vxlanorch.cpp:
```diff
@@ -185,6 +371,60 @@ create_tunnel_termination(
+        sai_ip_address_t ips, ipd, *ip=nullptr;
+        swss::copy(ips, src_ip_);
+
+        ids_.tunnel_decap_id = create_tunnel_map(decap);
```
no check decap for MAP_TO_INVALID?

pavel-shirshov commented on orchagent/vxlanorch.cpp:
```diff
@@ -220,24 +460,10 @@ bool VxlanTunnelOrch::addOperation(const Request& request)
-
-    SWSS_LOG_NOTICE("Vxlan tunnel '%s' was created", tunnel_name.c_str());
+    std::unique_ptr<VxlanTunnel> tunnel_obj (new VxlanTunnel(tunnel_name, src_ip, dst_ip));
+    vxlan_tunnel_table_[tunnel_name] = std::move(tunnel_obj);
```
why do we need this? Create an object and immediately move it into something?

pavel-shirshov commented on orchagent/vxlanorch.cpp:
```diff
@@ -258,7 +484,7 @@ bool VxlanTunnelMapOrch::addOperation(const Request& request)
     if(!gPortsOrch->getVlanByVlanId(vlan_id, tempPort))
     {
-        SWSS_LOG_ERROR("Vxlan tunnel map vlan id doesn't exist: %d", vlan_id);
+        SWSS_LOG_WARN("Vxlan tunnel map vlan id doesn't exist: %d", vlan_id);
```
I think this is error, whole operation failed.

pavel-shirshov commented on orchagent/vxlanorch.cpp:
```diff
@@ -273,23 +499,31 @@ bool VxlanTunnelMapOrch::addOperation(const Request& request)
     if (!tunnel_orch->isTunnelExists(tunnel_name))
     {
-        SWSS_LOG_ERROR("Vxlan tunnel '%s' doesn't exist", tunnel_name.c_str());
+        SWSS_LOG_WARN("Vxlan tunnel '%s' doesn't exist", tunnel_name.c_str());
```
whole operation failed, this is an error.

pavel-shirshov commented on orchagent/vxlanorch.cpp:
```diff
@@ -299,7 +533,8 @@ bool VxlanTunnelMapOrch::addOperation(const Request& request)
     }
 
-    SWSS_LOG_NOTICE("Vxlan tunnel map entry '%s' for tunnel '%s' was created", tunnel_map_entry_name.c_str(), tunnel_name.c_str());
+    SWSS_LOG_INFO("Vxlan tunnel map entry '%s' for tunnel '%s' was created",
```
Notice is better to see what object were created

pavel-shirshov commented on orchagent/vxlanorch.cpp:
```diff
@@ -312,3 +547,135 @@ bool VxlanTunnelMapOrch::delOperation(const Request& request)
+sai_object_id_t
+VxlanVrfMapOrch::createNextHopTunnel(const std::string& vName, IpAddress& ipAddr, MacAddress macAddress, uint32_t vni)
+{
+    if (vxlan_vrf_tunnel_.find(vName) ==  vxlan_vrf_tunnel_.end())
```
`==  vxlan` double space?

pavel-shirshov commented on orchagent/vxlanorch.cpp:
```diff
@@ -312,3 +547,135 @@ bool VxlanTunnelMapOrch::delOperation(const Request& request)
+        throw std::runtime_error("Can't create nexthop tunnel object");
+    }
+
+    SWSS_LOG_INFO("NH tunnel created for %s, id 0x%lx", vName.c_str(), nh_id);
```
`NH vxlan tunnel was created`

pavel-shirshov commented on orchagent/vxlanorch.cpp:
```diff
@@ -312,3 +547,135 @@ bool VxlanTunnelMapOrch::delOperation(const Request& request)
+
+    if (create_nexthop_tunnel(host_ip, vni, macptr, tunnel_id, &nh_id) != SAI_STATUS_SUCCESS)
+    {
+        throw std::runtime_error("Can't create nexthop tunnel object");
```
I think it's better to output host_ip, vni, mac

pavel-shirshov commented on orchagent/vxlanorch.cpp:
```diff
@@ -312,3 +547,135 @@ bool VxlanTunnelMapOrch::delOperation(const Request& request)
+    }
+    else
+    {
+        SWSS_LOG_WARN("VNet/Vrf not yet created %s", vrf_name.c_str());
```
Better `"VNet/Vrf '%s' hasn't been created yet"`

pavel-shirshov commented:
I see two functions with the same name, but with different signatures. If you need both functions I'd suggest to change names, otherwise you'd better to remove unused one.
```
sai_object_id_t create_tunnel_map_entry(
    MAP_T map_t,
    sai_object_id_t tunnel_map_id,
    sai_object_id_t obj_id,
    sai_uint32_t vni,
    bool encap)
```
vs
```
static sai_object_id_t
create_tunnel_map_entry(
    MAP_T map_t,
    sai_object_id_t tunnel_map_id,
    sai_uint32_t vni,
    sai_uint16_t vlan_id)
```

prsunny commented:
@pavel-shirshov , addressed the comment by combining the two APIs. Initially thought to leave the existing function unmodified. 

There are new changes to the PR based on VNET Table flow redesign. The new flow is updated in the [HLD ](https://github.com/Azure/SONiC/blob/gh-pages/doc/vxlan/Vxlan_hld.md) document.
